howso/attribute_maps.amlg

```diff
@@ -195,6 +195,202 @@
+			shared_deviations_map
+				(append
+					(map
+						(lambda
+							(or
+								(get shared_deviations_map (current_index))
+								[(current_index  1)]
+							)
+						)
+						(zip (apply "append" (values shared_deviations_map)))
+					)
+					shared_deviations_map
+				)
+		))
+
+		;create lists of referenced features for each shared feature
+		(declare (assoc
+			shared_sets
+				(map
+					(lambda (let
+						(assoc
+							shared_features (current_value 1)
+							feature (current_index 1)
+						)
+						(append
+							feature
+							;keep lists of shared features that reference either this feature or any of its shared_features
+							(indices (filter
+								(lambda
+									(or
+										(contains_value shared_features (current_index))
+										(contains_value (current_value) feature)
+										(!=
+											(size shared_features)
+											(size (remove (zip shared_features) (current_value)) )
+										)
+									)
+								)
+								(remove shared_deviations_map feature)
+							))
+						)
+					))
+					shared_deviations_map
+				)
+		))
+
+
+		(declare (assoc
+			combined_shared_sets
+				(map
+					(lambda (let
+						(assoc
+							shared_set (zip (current_value 1))
+							num_features_in_set (size (current_value 1))
+							feature (current_index 1)
+						)
+
+						;append all the overlapping shared sets and make them unique by zipping up the appended features
+						(zip
+							(apply "append"
+								;if there's any overlap between shared_set and any other feature's shared_set, combine them
+								(values (filter
+									(lambda
+										(if (!= num_features_in_set (size (remove shared_set (current_value))))
+											(append shared_set (zip (current_value)))
+										)
+									)
+									(remove shared_sets feature)
+								))
+							)
+						)
+
+					))
+					shared_sets
+				)
+
+		))
+
+		(declare (assoc
+			unique_sets (values combined_shared_sets (true))
+		))
+
+		(declare (assoc
+			shared_deviations_key_map
+				(map
+					(lambda (let
+						(assoc feature (current_index 1))
+						;find the the shared set that this feature belongs to and output the first feature from that set
+						(first (indices
+							(first (filter
+								(lambda (contains_index (current_value) feature))
+								unique_sets
+							))
+						))
+					))
+					combined_shared_sets
+				)
+		))
+
+		(assoc
+			"shared_deviations_key_map" shared_deviations_key_map
+			"shared_deviations_groupings" unique_sets
+		)
+	)
+
+	;Validates the necessary parameters for shared deviations, making sure that no shared deviations features are unique nominals.
+	#!ValidateSharedDeviations
+	(let
+		(assoc error_messages (list))
+		;check to see if any features in a shared deviations group is a unique nominal
+		(if (size unique_nominals_set)
+			(declare (assoc
+				invalid_unique_shared_deviations_features
+					(filter
+						(lambda
+							(contains_value (indices unique_nominals_set) (current_index))
```

Comment:
simplify `(contains_value (indices unique_nominals_set) (current_index))` to `(contains_index unique_nominals_set (current_index))